### PR TITLE
Add Sql.Collate extension/method to apply collation to string.

### PR DESCRIPTION
**Pull request by Corey-M**
Per Issue #2729

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/linq2db/linq2db?pullRequest=2731) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/linq2db/linq2db?pullRequest=2731) before we can accept your contribution.<br/><br/>:x: Corey-M [sign now](https://cla.dotnetfoundation.org/linq2db/linq2db?pullRequest=2731)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/linq2db/linq2db?pullRequest=2731) it.</sub>

**Review by igor-tkachev — CHANGES_REQUESTED** [review]
We also need unit test(s) for this feature.

**Review by MaceWindu — CHANGES_REQUESTED** [review]
If we add this to linq2db, we need a bit more code :)
1. generate proper collate clause for more databases (I suspect this one will not work for all dbs that support collate)
2. test
3. I would prefer to avoid regex use. Actually I don't think it even makes sense to validate collation - it's user responsibility and he will get error anyways (or SQL injection :rofl:  )

**Comment by MaceWindu** [comment]
Also as we don't enclose expression in brackets, we must specify proper operator precedence

**Comment by MaceWindu** [comment]
Made some documentation lookup:
- Access: no collation support
- DB2 LUW: COLLATION_KEY_BIT function https://www.ibm.com/support/knowledgecenter/en/SSEPGG_11.5.0/com.ibm.db2.luw.sql.ref.doc/doc/r0050491.html
- DB2 z/Os: COLLATION_KEY function(returns binary! also exists for DB2 LUW): https://www.ibm.com/support/knowledgecenter/SSEPEK_12.0.0/sqlref/src/tpc/db2z_bif_collationkey.html
- Firebird: standard "{str} COLLATE {collation}", e.g. https://www.firebirdsql.org/file/documentation/html/en/refdocs/fblangref25/firebird-25-language-reference.html#fblangref25-dml-select-column-list
- Informix: cannot find any dynamic collate support
- MYSQL/MariaDB: standard "{str} COLLATE {collation}" : https://dev.mysql.com/doc/refman/8.0/en/charset-collate.html, https://mariadb.com/kb/en/setting-character-sets-and-collations/#examples
- Oracle: standard "{str} COLLATE {collation}": https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/COLLATE-Operator.html#GUID-1B8CE3B0-77FC-455C-8400-6F81CF188D7B
- PostgreSQL: collation name is double-quoted (not string literal! which is single-quoted): https://www.postgresql.org/docs/9.1/collation.html
- SAP HANA: cannot find any dynamic collate support
- SQL CE: cannot find any dynamic collate support
- SQLITE: standard "{str} COLLATE {collation}": https://www.sqlite.org/datatype3.html#collation
- SQL Server: standard "{str} COLLATE {collation}": https://docs.microsoft.com/en-us/sql/t-sql/statements/collations?view=sql-server-ver15
- SYBASE ASE: cannot find any dynamic collate support

so basically only  following databases need special handling:
- postgresql
- DB2 LUW
- DB2 z/OS

**Review by MaceWindu — APPROVED** [review]

**Review by sdanyliv — APPROVED** [review]

**Review by viceroypenguin — APPROVED** [review]
